### PR TITLE
[2979] - add filtering for salary, qualification, vacancies & study type

### PR DESCRIPTION
**Pull request by TobyRet**
### Context

Adds filtering for salary, qualification, vacancies and study type.

### Changes proposed in this pull request
When filtering, calls can now be made to V3 API 😸 

### Guidance to review
- This is a big PR
- There is very little logic added to make the filtering work. However there are lots of changes in the specs. As discussed at the last tech brakedance we propose that specs should use a simpler method for stubbing JSON responses from the V3 API - i.e. stub at the network level rather than the JSON API resource layer.
- We have moved a lot of the specs out of the `results_spec` and put them into the filter specs. This is because feature specs should be testing the entire feature end to end, not just a page.

### CODE REVIEW

File: app/views/results/index.html.erb [hunk 1]
```diff
@@ -26,7 +26,7 @@
     <div class="govuk-grid-column-two-thirds">
       <div class="govuk-grid-row">
         <div class="govuk-grid-column-one-half">
-          <p class="govuk-body search-results__count"><%= number_with_delimiter(@courses.total_count) %> courses found</p>
+          <p class="govuk-body search-results__count"><%= number_with_delimiter(@results_view.total_course_count) %> courses found</p>
```
Comment (defong) [inline app/views/results/index.html.erb]: Zero courses issues

File: spec/fixtures/api_responses/courses.json [hunk 2]
```diff
@@ -0,0 +1,1065 @@
...
+          "contact_name": "Programme Director",
+          "year_code": "2019",
+          "provider_code": "E69",
+          "provider_type": "scitt",
+          "postcode": "EC3N 3AJ",
+          "website": "http://www.e-qualitas.org.uk",
+          "address1": "6th Floor, 10 Lloyd's Avenue",
+          "address2": "",
+          "address3": "London",
+          "email": "initialteachertraining@e-qualitas.co.uk",
```
Comment (defong) [inline spec/fixtures/api_responses/courses.json]: we shouldn't using email address that not ours

i fact shouldn't this be dummy data... not want to see this pop up in a google search

Comment (danielburnley) [reply spec/fixtures/api_responses/courses.json]: ^^ If we're using fixtures they should still be dummy data

File: spec/features/result_filters/location_spec.rb [hunk 3]
```diff
@@ -5,10 +5,35 @@
...
+        "page[page]" => 1,
+        "page[per_page]" => 10,
+    }
+  end
 
   before do
     stub_geocoder
-    stub_results_page_request
+
+    stub_api_v3_resource(
```
Comment (gpeng) [inline spec/features/result_filters/location_spec.rb]: I've just been having a look through this method. I think it bypasses the normal deserialization of the json and uses tests specific stuff. If that is what's happening I think for feature specs where we're trying to test the full stack we should stub the network response with webmock. That way we are testing the code in the model as well. We could use `stub_api_v3_request` for that I think but it takes a hash so we'd have the overhead of getting an example response in JSON and converting it. That's ok for small bits of data but in these specs the responses are quite big.

File: spec/fixtures/api_responses/courses.json (hunk 2 above)
Comment (gpeng) [reply spec/fixtures/api_responses/courses.json]: Sorry.. my fault. I thought the data was anonymised.

Comment (TobyRet) [reply spec/fixtures/api_responses/courses.json]: @defong @danielburnley 
QA, Staging and Prod data is not anonymised. It's public data - I’m not sure I see what the issue is?

File: spec/features/result_filters/location_spec.rb (hunk 3 above)
Comment (TobyRet) [reply spec/features/result_filters/location_spec.rb]: 👍

File: app/view_objects/results_view.rb [hunk 4]
```diff
@@ -141,4 +162,14 @@ def google_map_zoom
       "14"
     end
   end
+
+  def study_type
+    return "full_time,part_time" if fulltime? && parttime?
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/view_objects/results_view.rb]: Sorry not sure I understand @defong

File: app/views/results/index.html.erb [hunk 5]
```diff
@@ -1,9 +1,9 @@
-<%= content_for :page_title, "#{number_with_delimiter(@courses.total_count)} courses" %>
+<%= content_for :page_title, "#{number_with_delimiter(@results_view.total_course_count)} courses" %>
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/views/results/index.html.erb]: @defong - this has not been implemented yet. Should be done as part of https://trello.com/c/aHtGYeOY/2838-add-suggested-searches-to-location-searching

File: app/views/results/index.html.erb (hunk 1 above)
Comment (TobyRet) [reply app/views/results/index.html.erb]: Yep count needs to be fixed 👍

File: app/views/results/index.html.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (defong) [reply app/views/results/index.html.erb]: Page title... should be in another card altogether...

File: app/view_objects/results_view.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (defong) [reply app/view_objects/results_view.rb]: when both `full_time` and `part_time` are `false` than they are defaulted to be both `true`

File: spec/fixtures/api_responses/courses.json (hunk 2 above)
Comment (defong) [reply spec/fixtures/api_responses/courses.json]: We only need dummy data for the structure.

Everything else is excessive.

File: app/views/results/index.html.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/views/results/index.html.erb]: 👍

File: app/views/results/index.html.erb (hunk 1 above)
Comment (TobyRet) [reply app/views/results/index.html.erb]: https://github.com/DFE-Digital/find-teacher-training/pull/109/commits/712699839ae0ab984a8ad7897565bc97ff4bc0d7

File: app/view_objects/results_view.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/view_objects/results_view.rb]: Ok makes sense

[the remark this answers is not in the document]
Comment (TobyRet) [reply app/view_objects/results_view.rb]: https://github.com/DFE-Digital/find-teacher-training/pull/109/commits/084f52e026201957142785b7a0aa9cd90dc6b863

File: spec/fixtures/api_responses/courses.json (hunk 2 above)
Comment (TobyRet) [reply spec/fixtures/api_responses/courses.json]: Data has been anonymised

File: app/view_objects/results_view.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (gpeng) [reply app/view_objects/results_view.rb]: I don't think we need to do this. The Rails API returns all study modes by default so we don't need to pass a filter if we want all of them. We only need the filter if we want to restrict it to just full time or just part time.

[the remark this answers is not in the document]
Comment (TobyRet) [reply app/view_objects/results_view.rb]: OK - I can just drop this commit

[the remark this answers is not in the document]
Comment (defong) [reply app/view_objects/results_view.rb]: :+1:

**Pull request merged**